models/darknet/darknet_impl.hpp

```diff
@@ -0,0 +1,183 @@
+/**
+ * @file darknet_impl.hpp
+ * @author Kartik Dutt
+ *
+ * Implementation of LeNet using mlpack.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MODELS_DARKNET_IMPL_HPP
+#define MODELS_DARKNET_IMPL_HPP
+
+#include "darknet.hpp"
+
+namespace mlpack {
+namespace ann {
+
+template<
+     typename OutputLayerType,
+     typename InitializationRuleType,
+     size_t DarkNetVer
+>
+DarkNet<OutputLayerType, InitializationRuleType, DarkNetVer>::DarkNet() :
+    inputChannel(0),
+    inputWidth(0),
+    inputHeight(0),
+    numClasses(0),
+    weights("none")
+{
+  // Nothing to do here.
+}
+
+template<
+     typename OutputLayerType,
+     typename InitializationRuleType,
+     size_t DarkNetVer
+>
+DarkNet<OutputLayerType, InitializationRuleType, DarkNetVer>::DarkNet(
+  const size_t inputChannel,
+  const size_t inputWidth,
+  const size_t inputHeight,
+  const size_t numClasses,
+  const std::string& weights,
+  const bool includeTop) :
+  DarkNet<OutputLayerType, InitializationRuleType, DarkNetVer>(
+    std::tuple<size_t, size_t, size_t>(
+      inputChannel,
+      inputWidth,
+      inputHeight),
+      numClasses,
+      weights,
+      includeTop)
+{
+  // Nothing to do here.
+}
+
+template<
+     typename OutputLayerType,
+     typename InitializationRuleType,
+     size_t DarkNetVer
+>
+DarkNet<OutputLayerType, InitializationRuleType, DarkNetVer>::DarkNet(
+    const std::tuple<size_t, size_t, size_t> inputShape,
+    const size_t numClasses,
+    const std::string& weights,
+    const bool includeTop)
+{
+  mlpack::Log::Assert(DarkNetVer == 19 || DarkNetVer == 53,
+      "Incorrect DarkNet version. Possible values are 19 and 53. Trying \
+      to find version : " + std::to_string(DarkNetVer) + ".");
+
+  if (weights == "cifar10")
+  {
+    // Download weights here.
+    LoadModel("./../weights/darknet/darknet" + std::to_string(DarkNetVer) +
+        "_cifar10.bin");
+    return;
+  }
+  else if (weights != "none")
+  {
+    LoadModel(weights);
+    return;
+  }
+
+  if (DarkNetVer == 19)
+  {
+    darkNet.Add(new IdentityLayer<>());
+
+    // Convolution and activation function in a block.
+    darkNet.Add(ConvolutionBlock(inputChannel, 32, 3,
```

Comment:
So my approach would be to create the network layer by layer call the `Reset()` method and check the number of parameters `mode.Parameters().n_elem` just to check what the actual parameters are, maybe there is some uninitialized parameter which blows the model size.

Reply:
Sure, makes sense. I will try it and get back to you. Thanks a lot.

Reply:
So I think I figured out the issue a bit. This line [here](https://github.com/mlpack/models/pull/20/files#diff-ed328327354cde0c25612990454cf11dR134) is where the model blows out of proportion even if you add a single layer.

Reply:
Ahh, thanks for going through the process.